itest/lnd_custom_features.go

```diff
@@ -0,0 +1,92 @@
+package itest
+
+import (
+	"fmt"
+
+	"github.com/lightningnetwork/lnd/lnrpc"
+	"github.com/lightningnetwork/lnd/lnrpc/peersrpc"
+	"github.com/lightningnetwork/lnd/lntest"
+	"github.com/stretchr/testify/require"
+)
+
+// testCustomFeatures tests advertisement of custom features in various bolt 9
+// sets. For completeness, it also asserts that features aren't set in places
+// where they aren't intended to be.
+func testCustomFeatures(ht *lntest.HarnessTest) {
+	var (
+		// Odd custom features so that we don't need to worry about
+		// issues connecting to peers.
+		customInit    uint32 = 101
+		customNodeAnn uint32 = 103
+		customInvoice uint32 = 105
+	)
+
+	// Restart Alice with custom feature bits configured.
+	extraArgs := []string{
+		fmt.Sprintf("--protocol.custom-init=%v", customInit),
+		fmt.Sprintf("--protocol.custom-nodeann=%v", customNodeAnn),
+		fmt.Sprintf("--protocol.custom-invoice=%v", customInvoice),
+	}
+	ht.RestartNodeWithExtraArgs(ht.Alice, extraArgs)
+
+	// Connect nodes and open a channel so that Alice will be included
+	// in Bob's graph.
+	ht.ConnectNodes(ht.Alice, ht.Bob)
+	chanPoint := ht.OpenChannel(
+		ht.Alice, ht.Bob, lntest.OpenChannelParams{Amt: 1000000},
+	)
+
+	// Check that Alice's custom feature bit was sent to Bob in her init
+	// message.
+	peers := ht.Bob.RPC.ListPeers()
+	require.Len(ht, peers.Peers, 1)
+	require.Equal(ht, peers.Peers[0].PubKey, ht.Alice.PubKeyStr)
+
+	_, customInitSet := peers.Peers[0].Features[customInit]
+	require.True(ht, customInitSet)
+	assertNotSet(
+		ht, []uint32{customNodeAnn, customInvoice},
+		peers.Peers[0].Features,
+	)
+
+	// Assert that Alice's custom feature bit is contained in the node
+	// announcement sent to Bob.
+	updates := ht.AssertNumNodeAnns(ht.Bob, ht.Alice.PubKeyStr, 1)
```

Comment:
We'll need to do a length check on `updates` to avoid panic.

Reply:
wdym? Isn't `AssertNumNodeAnns` guaranteed to give me 1 update or fail